modules/roles/main.tf

```diff
@@ -0,0 +1,46 @@
+data "aws_caller_identity" "current" {}
+
+locals {
+  assume_role_policy = jsonencode(var.assume_role_policy != null ? var.assume_role_policy : {
```

Comment:
This one is tricky.. If we keep users (and Terraform user) on organization root account + Assume Role to switch to particular accununt, then `aws_caller_identity` will have current account_id (as it should..) BUT allowed principal should be root-account (as the only place where actual users are, but unknown at this context).
My suggestion is to change `var.assume_role_policy` to `var.allowed_principals` or something similar (maybe two dedicated vars?) to support both: users and AWS services to define the "Principal" part of this ARP. By default it MIGHT be current account, but I'd prefer no-default for sake of security. The module user will have to provide proper data as input explicitly so it's clear we can actually assume that role and elevate permissions.

Reply:
Yeah this should work nicely. It's internal module so it might be "less" user friendly - it requires you to know what you can put there.
To make it fully user friendly you have to provide 4 vars for all valid types of principals: AWS, Service, Federated and CanonicalUser and build this structure. You might consider using [aws_iam_policy_document](https://registry.terraform.io/providers/hashicorp/aws/latest/docs/data-sources/iam_policy_document) as it (should) provide future-proof and typo-free schema - check Assume-Role Policy with Multiple Principals example. 
If you pick this path (either manually building or with data resource), double check how it acts with empty lists! It might interpret empty list as ANYTHING, which is last thing we want ;)

Reply:
alright, so we would like sth like this?

```tf
variable "allowed_principals" {
  description = "object with allowed principals that grant permission to assume the role"
  type = object({
    AWS     = optional(set(string), [])
    Service = optional(set(string), [])
  })
}

data "aws_iam_policy_document" "assume_role_policy" {
  statement {
    actions = ["sts:AssumeRole"]
    dynamic "principals" {
      for_each = var.allowed_principals
      content {
        type        = principals.key
        identifiers = principals.value
      }
    }
  }
}

resource "aws_iam_role" "main" {
  force_detach_policies = true

  assume_role_policy = data.aws_iam_policy_document.assume_role_policy.json
  description        = var.description
  name               = var.name
  tags               = var.tags
}
```